File: starter-bots/PythonBot/Bot.py
from Action import Action
from PlayerCommand import PlayerCommand

import json

class Bot():
    def __init__(self) -> None:
        self.my_nodes = []
        self.my_scout_towers = []
        self.actions = []
        self.state = {}
        self.bot_id = 0

    def __str__(self) -> str:
        return "Bot Id:" + self.bot_id

    def set_bot_id(self, arg_id) -> None:
        self.bot_id = arg_id

    def set_bot_state(self, bot_state) -> None:
        self.state = bot_state
        self.my_scout_towers = bot_state.world.map.scoutTowers
        self.my_nodes = bot_state.world.map.nodes

    def __add_action(self, action) -> None:
        self.actions.append(action)

    def scout(self, units, node) -> None:
        self.__add_action(Action.scout_action(units, node))

    def farm(self, units, node) -> None:
        self.__add_action(Action.farm_action(units, node))

    def mine(self, units, node) -> None:
        self.__add_action(Action.mine_action(units, node))
    
    def lumber(self, units, node) -> None:
        self.__add_action(Action.lumber_action(units, node))

    def start_campfire(self, units, node) -> None:
        self.__add_action(Action.start_campfire_action(units, node))

    def get_command(self) -> PlayerCommand:
        return PlayerCommand(self.bot_id, self.actions)

    def toJson(self) -> str:
        return json.dumps(self, default=lambda o: o.__dict__)

    def __repr__(self) -> str:
        return self.toJson()

File: starter-bots/ReferenceBot/Domain/bot.js
const {scoutAction, farmAction, mineAction, lumberAction, startCampfireAction} = require('./actions');

class Bot {
    constructor() {
        this.configValues= {}
        
        this.myNodes = [];
        this.myScoutTowers = [];
        this.actions = [];
        this.state = {};
        this.baseLocation = null;
        this.food = 0;
        this.stone = 0;
        this.wood = 0;
        this.population = 0;
        this.availableUnits = 0;
        this.travellingUnits = 0;
        this.scoutingUnits = 0;
        this.farmingUnits = 0;
        this.lumberingUnits = 0;
        this.miningUnits = 0;
    }

    setBotId = (id) => {
        this.id = id;
    }

    setState = (botState) => {
        // console.log("================================================================================================");
        // console.log(JSON.stringify(botState, null, 4));        
        // console.log("================================================================================================");

        this.actions = [];
        const botDto = botState.bots.find(bot => bot.id === this.id);
        
        this.state = botState;
        this.myScoutTowers = botState.world.map.scoutTowers.filter(scoutTower => !botDto.map.scoutTowers.includes(scoutTower.id));
        this.myNodes = botState.world.map.nodes;
        this.currentTick = botState.world.currentTick;
        this.currentTier = botDto.currentTierLevel;
        this.baseLocation = botDto.baseLocation;
        this.availableUnits = botDto.availableUnits;
        this.travellingUnits = botDto.travellingUnits;
        this.scoutingUnits = botDto.scoutingUnits;
        this.farmingUnits = botDto.farmingUnits;
        this.lumberingUnits = botDto.lumberingUnits;
        this.miningUnits = botDto.miningUnits;
        this.population = botDto.population;
        this.food = botDto.food;
        this.stone = botDto.stone;
        this.wood = botDto.wood;
        this.heat = botDto.heat;
    }
    
    setConfigValues = (configValues) => {
        this.configValues = configValues;
    }

    // TODO: Check action type against resource type because you cant mine wood?
    #addAction = (action) => {
        this.actions.push(action);
        this.availableUnits -= action.Units;
    }
    scout = (units, node) => {
        console.log('Scouting [' + this.currentTick + ']');
        this.#addAction(scoutAction(units, node));
    }
    farm = (units, node) => {
        console.log('Farming [' + this.currentTick + ']');
        this.#addAction(farmAction(units, node));
    }
    mine = (units, node) => {
        console.log('Mining [' + this.currentTick + ']');
        this.#addAction(mineAction(units, node));
    }
    lumber = (units, node) => {
        console.log('Lumbering [' + this.currentTick + ']');
        this.#addAction(lumberAction(units, node));
    }
    startCampfire = (units) => {
        console.log('Starting Campfire[' + this.currentTick + ']');
        this.#addAction(startCampfireAction(units));
    }

    getCommand = () =>{
        return {
            PlayerId: this.id,
            Actions: [...this.actions]
        };
    }

    getMinimumResourcesForTier = () => {
        if(this.currentTier == 0){
            return {
                wood: 100,
                stone: 35,
                food: 20,
            }
        } else if(this.currentTier == 1){
            return {
                wood: 500,
                stone: 2500,
                food: 200,
            }
        } else if(this.currentTier == 2){
            return {
                wood: 5000,
                stone: 25000,
                food: 20000,
            }
        } else if(this.currentTier == 3){
            return {
                wood: 50000,
                stone: 250000,
                food: 20000,
            }
        } else if(this.currentTier == 4){
            return {
                wood: 500000,
                stone: 2500000,
                food: 200000,
            }
        } else {
            return {
                wood: Infinity,
                stone: Infinity,
                food: Infinity
            }
        }
    }
}

module.exports = {Bot};